README.md
# videojs-resize

Detects videojs player resize and adds/removes classes.

Resize detection inspired by [Daniel Buchner's post](http://www.backalleycoder.com/2013/03/18/cross-browser-event-based-element-resize-detection/).

## Installation

```sh
npm install --save videojs-resize
```

The npm installation is preferred, but Bower works, too.

```sh
bower install  --save videojs-resize
```

## Options
You can pass breakpoints as options to the plugin. These are the defaults.

The class is added if `player width < breakpoint key`.

```
player.resize({
  breakpoints: {
    300: 'vjs-size-tiny',
    400: 'vjs-size-small'
  }
});
```

## Usage

To include videojs-resize on your website or web application, use any of the following methods.

### `<script>` Tag

This is the simplest case. Get the script in whatever way you prefer and include the plugin _after_ you include [video.js][videojs], so that the `videojs` global is available.

```html
<script src="//path/to/video.min.js"></script>
<script src="//path/to/videojs-resize.min.js"></script>
<script>
  var player = videojs('my-video');

  player.resize();
</script>
```

### Browserify

When using with Browserify, install videojs-resize via npm and `require` the plugin as you would any other module.

```js
var videojs = require('video.js');

// The actual plugin function is exported by this module, but it is also
// attached to the `Player.prototype`; so, there is no need to assign it
// to a variable.
require('videojs-resize');

var player = videojs('my-video');

player.resize();
```

### RequireJS/AMD

When using with RequireJS (or another AMD library), get the script in whatever way you prefer and `require` the plugin as you normally would:

```js
require(['video.js', 'videojs-resize'], function(videojs) {
  var player = videojs('my-video');

  player.resize();
});
```


CHANGELOG.md
CHANGELOG
=========

## HEAD
* 1.0.0

--------------------



test/karma.conf.js
module.exports = function(config) {
  var detectBrowsers = {
    enabled: false,
    usePhantomJS: false
  };

  // On Travis CI, we can only run in Firefox.
  if (process.env.TRAVIS) {
    config.browsers = ['Firefox'];
  }

  // If no browsers are specified, we enable `karma-detect-browsers`
  // this will detect all browsers that are available for testing
  if (!config.browsers.length) {
    detectBrowsers.enabled = true;
  }

  config.set({
    basePath: '..',
    frameworks: ['qunit', 'detectBrowsers'],

    files: [
      'dist/videojs-resize.css',
      'node_modules/sinon/pkg/sinon.js',
      'node_modules/sinon/pkg/sinon-ie.js',
      'node_modules/video.js/dist/video.js',
      'node_modules/video.js/dist/video-js.css',
      'test/dist/bundle.js'
    ],

    detectBrowsers: detectBrowsers,
    reporters: ['dots'],
    port: 9876,
    colors: true,
    autoWatch: false,
    singleRun: true,
    concurrency: Infinity
  });
};


test/plugin.test.js
import document from 'global/document';

import QUnit from 'qunit';
import sinon from 'sinon';
import videojs from 'video.js';

import plugin from '../src/plugin';

const Player = videojs.getComponent('Player');

QUnit.test('the environment is sane', function(assert) {
  assert.strictEqual(typeof Array.isArray, 'function', 'es5 exists');
  assert.strictEqual(typeof sinon, 'object', 'sinon exists');
  assert.strictEqual(typeof videojs, 'function', 'videojs exists');
  assert.strictEqual(typeof plugin, 'function', 'plugin is a function');
});

QUnit.module('videojs-resize', {

  beforeEach() {

    // Mock the environment's timers because certain things - particularly
    // player readiness - are asynchronous in video.js 5. This MUST come
    // before any player is created; otherwise, timers could get created
    // with the actual timer methods!
    this.clock = sinon.useFakeTimers();

    this.fixture = document.getElementById('qunit-fixture');
    this.video = document.createElement('video');
    this.fixture.appendChild(this.video);
    this.player = videojs(this.video);
  },

  afterEach() {
    this.player.dispose();
    this.clock.restore();
  }
});

QUnit.test('registers itself with video.js', function(assert) {
  assert.expect(2);

  assert.strictEqual(
    Player.prototype.resize,
    plugin,
    'videojs-resize plugin was registered'
  );

  this.player.resize();

  // Tick the clock forward enough to trigger the player to be "ready".
  this.clock.tick(1);

  assert.ok(
    this.player.hasClass('vjs-resize'),
    'the plugin adds a class to the player'
  );
});


src/plugin.js
import videojs from 'video.js';
import throttle from 'lodash-compat/function/throttle';

// Default options for the plugin.
const defaults = {
  breakpoints: {
    300: 'vjs-size-tiny',
    400: 'vjs-size-small'
  }
};

class Resize {
  constructor(player, options) {
    this.player = player;
    this.options = options;
    this.classes = [];
    this.widths = [];
  }

  setup() {
    const iframe = document.createElement('iframe');
    const html = `
    <script>
      window.addEventListener("resize", function() {
        parent.postMessage("resize:${this.player.id()}", "*");
      });
    </script>
    `;

    iframe.id = 'sizeframe';
    this.player.el().appendChild(iframe);
    iframe.contentWindow.document.open();
    iframe.contentWindow.document.write(html);
    iframe.contentWindow.document.close();

    Object.keys(this.options.breakpoints).map((key, index) => {
      this.classes.push(this.options.breakpoints[key]);
      this.widths.push(key);
    });

    this.player.addClass('vjs-resize');

    this.handleResize = this.onResize.bind(this);
    window.addEventListener('message', throttle(this.handleResize, 150));
  }

  removeClasses() {
    this.classes.forEach((item) => {
      this.player.removeClass(item);
    });
  }

  onResize(event) {
    let playerWidth = this.player.el().offsetWidth;
    let id = event.data.replace('resize:', '');

    // confirm message is from the current player
    if (event.data.indexOf('resize:') === 0 && id === this.player.id()) {
      this.removeClasses();

      for (let i = 0; i < this.widths.length; i++) {
        let width = this.widths[i];

        if (playerWidth < width) {
          this.player.addClass(this.options.breakpoints[width]);
        }
      }
    }
  }
}

/**
 * A video.js plugin.
 *
 * In the plugin function, the value of `this` is a video.js `Player`
 * instance. You cannot rely on the player being in a "ready" state here,
 * depending on how the plugin is invoked. This may or may not be important
 * to you; if not, remove the wait for "ready"!
 *
 * @function resize
 * @param    {Object} [options={}]
 *           An object of options left to the plugin author to define.
 */
const resize = function(options) {
  this.ready(() => {
    const resizeInstance = new Resize(this, videojs.mergeOptions(defaults, options));

    resizeInstance.setup();
  });
};

// Register the plugin with video.js.
videojs.plugin('resize', resize);

// Include the version number.
resize.VERSION = '__VERSION__';

export default resize;
